### ISSUE TITLE
Postponed RE with split()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6344]
Migrated from [rt.perl.org#21411](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=21411) (status was 'resolved')

Searchable as RT21411$

**Comment by p5pRT** [comment]
### From book@cpan.org

#### Created by book@cpan.org



I wanted to try the experimental \(??\{ \}\) regexp construct in a split&#8203;:

\#\!/usr/bin/perl
\#use re 'debug';
my $i; sub n \{ \+\+$i \}
my $c = 'a';
my $s = join ''\, map \{ $c\+\+ x $\_ \} 1 \.\. shift;
print join "|"\, split /\(\(??\{"\."x n\(\)\}\)\)/\, $s;

$ perl split\_with\_postponed\_re 1
|a

$ perl split\_with\_postponed\_re 2
|a||bb

$ perl split\_with\_postponed\_re 3
|a||\.\.\.||ccc
Expected&#8203;: |a||bb||ccc

$ perl split\_with\_postponed\_re 4
|a||\.\.\.\.|4|ccc||dddd
Expected&#8203;: |a||bb||ccc||dddd

$ perl split\_with\_postponed\_re 5
|a||\.\.\.\.\.|5|ccc||dddd||eeeee
Expected&#8203;: |a||bb||ccc||dddd||eeeee

Commenting out the use re 'debug' line gives a rather long output\, but
it looks quite normal\.

Tested with perl5\.8\.0\, perl5\.6\.1 and bleadperl\, with the same results\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by buildd at Sat Nov 30 17:45:17 UTC 2002.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i386-linux
    uname='linux cyberhq 2.4.19 #1 smp sun aug 4 11:30:45 pdt 2002 i686 unknown '
    config_args='-Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.6.1 -Darchlib=/usr/lib/perl/5.6.1 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.6.1 -Dsitearch=/usr/local/lib/perl/5.6.1 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Duseshrplib -Dlibperl=libperl.so.5.6.1 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011002 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -ldb -ldl -lm -lc -lcrypt
    perllibs=-ldl -lm -lc -lcrypt
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=true, libperl=libperl.so.5.6.1
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /home/book/sources/perl
    /home/book/bin/perl
    /usr/local/lib/perl/5.6.1
    /usr/local/share/perl/5.6.1
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.6.1
    /usr/share/perl/5.6.1
    /usr/local/lib/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/book
    LANG=fr_FR@euro
    LANGUAGE (unset)
    LC_ALL=french
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/book/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/usr/games:/usr/local/games:/usr/local/lib/jdk/bin:.
    PERL5LIB=/home/book/sources/perl:/home/book/bin/perl
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Sat\, Mar 01\, 2003 at 12&#8203;:16&#8203;:51AM \-0000\, Philippe 'BooK' Bruhat \(via RT\) wrote&#8203;:
> I wanted to try the experimental \(??\{ \}\) regexp construct in a split&#8203;:
> 
> \#\!/usr/bin/perl
> \#use re 'debug';
> my $i; sub n \{ \+\+$i \}
> my $c = 'a';
> my $s = join ''\, map \{ $c\+\+ x $\_ \} 1 \.\. shift;
> print join "|"\, split /\(\(??\{"\."x n\(\)\}\)\)/\, $s;
\.\.\.
> $ perl split\_with\_postponed\_re 3
> |a||\.\.\.||ccc
> Expected&#8203;: |a||bb||ccc

Simpler example&#8203;:
$ perl \-le 'print join "|"\, split /\(\(??\{$i\+\+\}\)\)/\,"\-1\-3\-5\-"'

This should print '\-|1|\-|3|\-|5|\-' not '\-|1|\-|7|\-|5|\-'

The expression inside \(??\{ \.\.\. \}\) corrupts the stack\.

Regards

Adi

Test for t/op/pat\.t&#8203;:


---
\{
&nbsp;    my $i;
&nbsp;    ok\('\-1\-3\-5\-' eq join\(''\, split /\(\(??\{$i\+\+\}\)\)/\, '\-1\-3\-5\-'\)\,
&nbsp;    "\[perl \#21411\] \(??\{ \.\. \}\) corrupts split's stack"\)
\}


---
Patch&#8203;:


---
<details><summary>Inline Patch</summary>

```diff
--- /arc/perl-current/pp.c	2003-02-25 16:35:04.000000000 +0200
+++ pp.c	2003-03-02 09:47:57.000000000 +0200
@@ -4632,6 +4632,7 @@ PP(pp_split)
 		}
 	    }
 	    s = rx->endp[0] + orig;
+	    PUTBACK;
 	}
     }
 
```
</details>

**Comment by p5pRT** [comment]
### From @hvds

Enache Adrian \<enache@&#8203;rdslink\.ro> wrote&#8203;:
:Simpler example&#8203;:
:$ perl \-le 'print join "|"\, split /\(\(??\{$i\+\+\}\)\)/\,"\-1\-3\-5\-"'
:
:This should print '\-|1|\-|3|\-|5|\-' not '\-|1|\-|7|\-|5|\-'
:
:The expression inside \(??\{ \.\.\. \}\) corrupts the stack\.

Thanks\, patch applied as \#18805\.

Hugo

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Sun\, Mar 02\, 2003 at 05&#8203;:41&#8203;:38PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
> Enache Adrian \<enache@&#8203;rdslink\.ro> wrote&#8203;:
> :Simpler example&#8203;:
> :$ perl \-le 'print join "|"\, split /\(\(??\{$i\+\+\}\)\)/\,"\-1\-3\-5\-"'
> :
> :This should print '\-|1|\-|3|\-|5|\-' not '\-|1|\-|7|\-|5|\-'
> :
> :The expression inside \(??\{ \.\.\. \}\) corrupts the stack\.
> 
> Thanks\, patch applied as \#18805\.

$ perl \-le 'split /\(?\{"FOO"\}\)/\,"abcde";print @&#8203;\_'
SEGV

That's split's stack swapping when the list it returns has to be
assigned to an array \(@&#8203;\_ here\) \. See \[perl \#20912\] too\.

In an expression like&#8203;:
&nbsp;  @&#8203;a = split /\(?\{ \.\. CODE \.\. \}\) \.\. /\, STRING;
the CODE will just use @&#8203;a as its stack and may try to reallocate it
\( this may happen with the regular Perl stack too\, of course \)\.

A q\-n\-d fix is to put a SPAGAIN too there to keep PL\_stack\_base and
SP referring to the same array\. \( see pp\.c&#8203;:4700 \)
It will keep the coredump off\, but certainly isn't a long term
solution\.


---
<details><summary>Inline Patch</summary>

```diff
--- /arc/perl-current/pp.c	2003-03-03 08:45:50.000000000 +0200
+++ pp.c	2003-03-04 13:25:43.000000000 +0200
@@ -4647,6 +4647,7 @@ PP(pp_split)
     }
     else {
 	maxiters += slen * rx->nparens;
+	PUTBACK;
 	while (s < strend && --limit
 /*	       && (!rx->check_substr
 		   || ((s = CALLREG_INTUIT_START(aTHX_ rx, sv, s, strend,
@@ -4654,6 +4655,7 @@ PP(pp_split)
 */	       && CALLREGEXEC(aTHX_ rx, s, strend, orig,
 			      1 /* minend */, sv, NULL, 0))
 	{
+	    SPAGAIN;
 	    TAINT_IF(RX_MATCH_TAINTED(rx));
 	    if (RX_MATCH_COPIED(rx) && rx->subbeg != orig) {
 		m = s;
-------------------------------------------------------------------


```
</details>
Probably the right place to fix that is the EVAL case in S_regmatch.
What about a special stack for the '\(?\{ \.\.\. \}\)' blocks to SWITCHSTACK
to it and then back there ?

Regards
Adi

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Tue\, Mar 04\, 2003 at 02&#8203;:20&#8203;:15PM \+0200\, Enache Adrian wrote&#8203;:
> On Sun\, Mar 02\, 2003 at 05&#8203;:41&#8203;:38PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
> > Enache Adrian \<enache@&#8203;rdslink\.ro> wrote&#8203;:
> > :Simpler example&#8203;:
> > :$ perl \-le 'print join "|"\, split /\(\(??\{$i\+\+\}\)\)/\,"\-1\-3\-5\-"'
> > :
> > :This should print '\-|1|\-|3|\-|5|\-' not '\-|1|\-|7|\-|5|\-'
> > :
> > :The expression inside \(??\{ \.\.\. \}\) corrupts the stack\.
> > 
> > Thanks\, patch applied as \#18805\.
> 
> $ perl \-le 'split /\(?\{"FOO"\}\)/\,"abcde";print @&#8203;\_'
> SEGV

A patch which fixes the above if a more nice \(IMHO\) way
\( without stack switching though \) follows\. \- The previous
patch left a chance for the stack to be reallocated during
the last\, failed match\.

It also fixes another little nasty bug&#8203;:

$ perl \-e 'split /\(?\{ split "" \}/\,"abc"'
SEGV

It has to do with regexp caching \- that's only PL\_regdata which
causes problems\, but I think it's better to restore them all
\(PL\_regnpar\, PL\_regprecomp\, etc\)\.

Regards
Adi



---
<details><summary>Inline Patch</summary>

```diff
--- /arc/perl-current/regexec.c	2003-03-02 18:34:00.000000000 +0200
+++ regexec.c	2003-03-06 17:34:07.000000000 +0200
@@ -2838,6 +2838,7 @@ S_regmatch(pTHX_ regnode *prog)
 	    COP *ocurcop = PL_curcop;
 	    PAD *old_comppad;
 	    SV *ret;
+	    struct regexp *oreg = PL_reg_re;
 	
 	    n = ARG(scan);
 	    PL_op = (OP_4tree*)PL_regdata->data[n];
@@ -2970,8 +2971,10 @@ S_regmatch(pTHX_ regnode *prog)
 		sw = SvTRUE(ret);
 		logical = 0;
 	    }
-	    else
+	    else {
 		sv_setsv(save_scalar(PL_replgv), ret);
+		cache_re(oreg);
+	    }
 	    break;
 	}
 	case OPEN:
--- /arc/perl-current/pp.c	2003-03-03 08:45:50.000000000 +0200
+++ pp.c	2003-03-06 18:53:05.000000000 +0200
@@ -4647,13 +4647,13 @@ PP(pp_split)
     }
     else {
 	maxiters += slen * rx->nparens;
-	while (s < strend && --limit
-/*	       && (!rx->check_substr
-		   || ((s = CALLREG_INTUIT_START(aTHX_ rx, sv, s, strend,
-						 0, NULL))))
-*/	       && CALLREGEXEC(aTHX_ rx, s, strend, orig,
-			      1 /* minend */, sv, NULL, 0))
+	while (s < strend && --limit)
 	{
+	    PUTBACK;
+	    i = CALLREGEXEC(aTHX_ rx, s, strend, orig, 1 , sv, NULL, 0);
+	    SPAGAIN;
+	    if (i == 0)
+		break;
 	    TAINT_IF(RX_MATCH_TAINTED(rx));
 	    if (RX_MATCH_COPIED(rx) && rx->subbeg != orig) {
 		m = s;
@@ -4692,7 +4692,6 @@ PP(pp_split)
 		}
 	    }
 	    s = rx->endp[0] + orig;
-	    PUTBACK;
 	}
     }
 
--- /arc/perl-current/t/op/pat.t	2003-03-02 19:39:51.000000000 +0200
+++ t/op/pat.t	2003-03-06 22:04:16.000000000 +0200
@@ -6,7 +6,7 @@
 
 $| = 1;
 
-print "1..994\n";
+print "1..996\n";
 
 BEGIN {
     chdir 't' if -d 't';
@@ -3142,8 +3142,16 @@ ok("bbbbac" =~ /$pattern/ && $1 eq 'a', 
 {
     my $i;
     ok('-1-3-5-' eq join('', split /((??{$i++}))/, '-1-3-5-'),
-	"[perl #21411] (??{ .. }) corrupts split's stack")
+	"[perl #21411] (??{ .. }) corrupts split's stack");
+    split /(?{'WOW'})/, 'abc';
+    ok('a|b|c' eq join ('|', @_),
+	"[perl #21411] (?{ .. }) version of the above");
 }
 
-# last test 994
+{
+    split /(?{ split "" })/, "abc";
+    ok(1,'cache_re & "(?{": it dumps core in 5.6.1 & 5.8.0');
+}
+
+# last test 996
 
```
</details>

**Comment by p5pRT** [comment]
@cwest - Status changed from 'new' to 'resolved'
